uid-mux/src/yamux.rs

```diff
@@ -0,0 +1,440 @@
+//! Yamux multiplexer.
+//!
+//! This module provides a [`yamux`](https://crates.io/crates/yamux) wrapper which implements [`UidMux`](crate::UidMux).
+
+use std::{
+    collections::HashMap,
+    fmt,
+    future::IntoFuture,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+    task::{Context, Poll, Waker},
+};
+
+use async_trait::async_trait;
+use futures::{
+    stream::FuturesUnordered, task::noop_waker, AsyncRead, AsyncWrite, Future, FutureExt, StreamExt,
+};
+use tokio::sync::{oneshot, Notify};
+use yamux::Connection;
+
+use crate::{
+    future::{ReadId, ReturnStream},
+    log::{debug, error, info, trace},
+    InternalId, UidMux,
+};
+
+pub use yamux::{Config, ConnectionError, Mode, Stream};
+
+type Result<T, E = ConnectionError> = std::result::Result<T, E>;
+
+#[derive(Debug, Clone, Copy)]
+enum Role {
+    Client,
+    Server,
+}
+
+impl fmt::Display for Role {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Role::Client => write!(f, "Client"),
+            Role::Server => write!(f, "Server"),
+        }
+    }
+}
+
+/// A yamux multiplexer.
+#[derive(Debug)]
+pub struct Yamux<Io> {
+    role: Role,
+    conn: Connection<Io>,
+    queue: Arc<Mutex<Queue>>,
+    close_notify: Arc<Notify>,
+    shutdown_notify: Arc<AtomicBool>,
+}
+
+#[derive(Debug)]
+struct Queue {
+    waiting: HashMap<InternalId, oneshot::Sender<Stream>>,
+    ready: HashMap<InternalId, Stream>,
+    waker: Waker,
```

Comment:
In my mind the waker causes the future to be polled, so it's more fitting to be a field in `YamuxFuture`.
I have a hard time mapping waker to the `Queue`.
In the code we wake the Future in order to close the connection, regardless if there is anything in the Queue, so i don't see a direct connection between waker and `Queue`.